### ISSUE TITLE
CONSIDER describing `tb` as out-of-scope (rather than not permitted) in this PEP.

### ISSUE THREAD

**Issue by davepeck** [davepeck/peps#31]
[As suggested by Chris Markiewicz](https://discuss.python.org/t/pep750-template-strings-new-updates/71594/4). (aka @effigies)

As he points out, [the reason f-strings don't combine with `b`](https://peps.python.org/pep-0498/#no-binary-f-strings) is because an object's `__format__()` may return unicode incompatible with the byte string.

Chris is right that, in theory, this isn't an issue with t-strings, since processing code can independently decide how to interpret interpolation format specs.

However, a lot of Template processing code will probably invoke `format(interpolation.value, interpolation.format_spec)` to do its work, and `format()`, of course, may invoke the `__format__` dunder. It might just be awkward to allow `tb` in practice, even in the future?

**Comment by davepeck** [comment]
After discussion @jimbaker and I agree that this is a good catch! We should keep the door open for `tb` in the future. One could even imagine using the byte packing mini-language in `struct` as a format spec.

**Comment by effigies** [comment]
> a lot of Template processing code will probably invoke `format(interpolation.value, interpolation.format_spec)` to do its work

Just as functions that expect `str` will fail if passed `bytes`, and vice versa, I would expect functions that expect `Template[str]` will fail if passed `Template[bytes]`, and type checkers could very easily warn in these cases.

If/when `Template[bytes]` is defined, I think it's entirely plausible that `Interpolation` contents will still be a `str`, since it's Python source code, so it could be that the only needed change is:

```python
class Template[T: str | bytes]:
    args: tuple[T | Interpolation, ...]

    def __init__(self, *args: T | Interpolation):
        ...

    @property
    def strings(self) -> tuple[T, ...]:
        ...

    @property
    def interpolations(self) -> tuple[Interpolation, ...]:
        ...

    @property
    def values(self) -> tuple[object, ...]:
        ...
```

If that's the case, you might consider using `.literals` instead of `.strings` for potential forward compatibility:

```diff
     @property
-    def strings(self) -> tuple[T, ...]:
+    def literals(self) -> tuple[T, ...]:

```

If you also need `Interpolation[bytes]` instead of `Interpolation[str]`, that might be more painful. But looking at it, I can't see why you'd want `expr` or `format_spec` to change types, and if anything you'd probably just want to forbid `conv`.

**Comment by davepeck** [comment]
> Just as functions that expect `str` will fail if passed `bytes`, and vice versa, I would expect functions that expect `Template[str]` will fail if passed `Template[bytes]`, and type checkers could very easily warn in these cases.

Yes, agreed.

> If/when `Template[bytes]` is defined, I think it's entirely plausible that `Interpolation` contents will still be a `str`, since it's Python source code, so it could be that the only needed change is:

Yes, that seems right: only `Template` is impacted.

Thanks for the suggestion here!

**Comment by davepeck** [comment]
Closed with `f4b03d62`
